Pycloud Deployer/infra_classes.py

```diff
@@ -0,0 +1,629 @@
+from config import logger, location, subscription_id
+from azure.mgmt.resource import ResourceManagementClient
+from azure.mgmt.network import NetworkManagementClient
+from azure.mgmt.compute import ComputeManagementClient
+from modules import (
+    create_publicip_params, 
+    create_nic_params,
+    create_server_params, 
+    run_command_params,
+    validate_group_name
+)
+from azure.core.exceptions import (
+    ResourceExistsError, 
+    HttpResponseError, 
+    ResourceNotFoundError
+)
+
+
+class Infrastructure:
+    """Set up the support infrastructure and deploy the resources, retrieve
+    information about the deployment and tear down a deployment
+
+    Methods:
+        create_infrastructure(): Create the support infrastructure.
+        info(): Display infrastructure information
+        delete(): Delete a deployment
+        tier2(): Deploy a tier 2 infrastructure of webservers and dbservers
+        tier3(): Deploy a 3-tier infrastructure of webservers, dbservers and load balancers
+        create_resource_group(): Create resource group
+        create_nsg_group(): Create NSG group
+        create_vnet(): Create a VNet
+        create_subnet(): Create a subnet
+    """
+    def __init__(self, token_credential):
+        """Initiate the class properties
+
+        Args:
+            token_credential (DefaultAzureCredential): Instance of the DefaultAzureCredential class
+        """
+        self.rss_client = ResourceManagementClient(token_credential, subscription_id)
+        self.net_client = NetworkManagementClient(token_credential, subscription_id)
+        self.compute_client = ComputeManagementClient(token_credential, subscription_id)
+
+        self.webserver_provisioner = Server(token_credential)
+        self.dbserver_provisioner = DBServer(token_credential)
+        self.lb_provisioner = LoadBalancer(token_credential)
+    
+    def create_infrastructure(self, group_name, vnet_name, subnet_name, 
+                              nsg_name, isTier2):
+        """Create the support infrastructure
+
+        Args:
+            group_name (str): Name of the resource group
+            vnet_name (str): Name of the VNet
+            subnet_name (str): Name of the subnet
+            nsg_name (str): Name of the NSG group
+        """
+        self.create_vnet(vnet_name, group_name)
+
+        nsg_info = self.create_nsg_group(nsg_name, group_name, isTier2)
+
+        # Likely error if there was a problem deploying the NSG and nothing was 
+        # returned to the nsg_info variable
+        try:
+            self.create_subnet(subnet_name, group_name, vnet_name, nsg_info.id)
+        except UnboundLocalError as e:
```

Comment:
Catching UnboundLocalError in create_infrastructure() and other methods is not a common practice and could be a sign of a design issue. It would be better to check if nsg_info is None or not before using it, rather than catching an error when it's not set:

something like

if nsg_info is not None:
    self.create_subnet(subnet_name, group_name, vnet_name, nsg_info.id)
else:
    logger.error("NSG information is missing, cannot create subnet.")

could be more readable